pinot-common/src/main/java/org/apache/pinot/common/request/context/LiteralContext.java

```diff
@@ -37,68 +45,116 @@ public class LiteralContext {
+        return ImmutablePair.of(FieldSpec.DataType.BIG_DECIMAL, new BigDecimal(literal));
+      } else {
+        return ImmutablePair.of(FieldSpec.DataType.STRING, literal);
+      }
+    } catch (Exception e) {
+      // Ignored
+    }
+
+    // Try to interpret the literal as TIMESTAMP
+    try {
+      Timestamp timestamp = Timestamp.valueOf(literal);
+      return ImmutablePair.of(FieldSpec.DataType.TIMESTAMP, timestamp);
+    } catch (Exception e) {
+      // Ignored
     }
+    return ImmutablePair.of(FieldSpec.DataType.STRING, literal);
   }
 
   public LiteralContext(Literal literal) {
-    _type = convertThriftTypeToDataType(literal.getSetField());
-    _value = literal.getFieldValue();
+    Preconditions.checkState(literal.getFieldValue() != null,
+        "Field value cannot be null for field:" + literal.getSetField());
+    switch (literal.getSetField()){
+      case BOOL_VALUE:
+        _type = FieldSpec.DataType.BOOLEAN;
+        _value = literal.getFieldValue();
+        break;
+      case DOUBLE_VALUE:
+        _type = FieldSpec.DataType.DOUBLE;
+        _value = literal.getFieldValue();
+        break;
+      case LONG_VALUE:
+        _type = FieldSpec.DataType.LONG;
+        _value = literal.getFieldValue();
+        break;
+      case NULL_VALUE:
+        _type = FieldSpec.DataType.UNKNOWN;
+        _value = null;
+        break;
+      case STRING_VALUE:
+        Pair<FieldSpec.DataType, Object> typeAndValue = inferLiteralDataTypeAndValue(literal.getFieldValue().toString());
+        _type = typeAndValue.getLeft();
+        _value = typeAndValue.getRight();
+        break;
+      default:
+        throw new UnsupportedOperationException("Unsupported data type:" + literal.getSetField());
+    }
+    _bigDecimalValue = getBigDecimalValue(_type, _value);
   }
 
   public FieldSpec.DataType getType() {
     return _type;
   }
 
+  public int getIntValue(){
+    return _bigDecimalValue.intValue();
+  }
+  public double getDoubleValue(){
+    return _bigDecimalValue.doubleValue();
+  }
+  public BigDecimal getBigDecimalValue() {
+    return _bigDecimalValue;
+  }
+
+  public String getStringValue() {
+    return String.valueOf(_value);
+  }
+
   @Nullable
   public Object getValue() {
     return _value;
   }
 
+  // This ctor is only used for special handling in subquery.
   public LiteralContext(FieldSpec.DataType type, Object value) {
-    Preconditions.checkArgument(convertDataTypeToJavaType(type) == value.getClass(),
-        "Unmatched data type: " + type + " java type: " + value.getClass());
     _type = type;
-    _value = value;
+    if(type == FieldSpec.DataType.UNKNOWN){
+      _value = null;
+    } else {
+      _value = value;
+    }
+    _bigDecimalValue = getBigDecimalValue(type, value);
   }
 
   @Override
   public int hashCode() {
-    if (_value == null) {
-      return 31 * _type.hashCode();
-    }
-    return 31 * _value.hashCode();
+    return 31 * Objects.hashCode(_value) + Objects.hashCode(_type);
```

Comment:
Why this change in the hash code?

Reply:
because originally we are not table to tell "123" and 123. After the type reservation, "123" and 123 are different. thus we add the type encoding into the hashcode.

Reply:
> because originally we are not table to tell "123" and 123. After the type reservation

Yeah, but I mean... why do we add the hash of types? That isn't usual when producing hashes. In fact if you do `return Objects.hashCode(_value, _type)` it will produce better hashCode (at the cost of creating a short living array).

Reply:
good point. changed to Objects.hash(...)